src/context/user_context.js

```diff
@@ -0,0 +1,114 @@
+    async function fetchData() {
+      const url = `${rootUrl}/api/v1/users/showMe`;
+      axios
+        .get(url, { withCredentials: true })
+        .then((response) => {
+          console.log(response);
+          setCurrentUser(response.data.user);
+        })
+        .catch((error) => {
+          const errorPayload =
+            error instanceof AxiosError ? error.response.data : error;
+          console.error(errorPayload);
+          setAuthError(errorPayload);
+        });
+    }
+    fetchData();
+    // By using empty array [], for the "dependencies" argument of useEffect, it tells React to run this useEffect hook only *once*, the first time this component/context is rendered
+  }, []);
+
+  const handleLogin = async (userCredentials) => {
+    setAuthLoading(true);
+    try {
+      const url = `${rootUrl}/api/v1/auth/login`;
+      // We could use fetch or axios. Fetch is native to the browser (doesn't need extra libs), but since we already have axios in use in this repo, I'll stick to using that. It's a little bit cleaner too than raw fetch, but either one is fine.
+      const response = await axios.post(url, userCredentials, {
+        withCredentials: true,
+      });
+      console.log(response);
+      setAuthLoading(false);
+      setCurrentUser(response.data.user);
+      return true;
+    } catch (error) {
+      console.error(error);
+      const errorPayload =
+        error instanceof AxiosError ? error.response.data : error;
+      setAuthLoading(false);
+      setAuthError(errorPayload);
+      return false;
+    }
+  };
+
+  const handleRegister = async (userCredentials) => {
+    setAuthLoading(true);
+    try {
+      const url = `${rootUrl}/api/v1/auth/register`;
+      const response = await axios.post(url, userCredentials, {
+        withCredentials: true,
+      });
+      console.log(response);
+      setCurrentUser(response.data.user);
+      setAuthLoading(false);
+      return true;
+    } catch (error) {
+      console.error(error);
+      const errorPayload =
+        error instanceof AxiosError ? error.response.data : error;
+      setAuthError(errorPayload);
+      setAuthLoading(false);
+      return false;
+    }
+  };
+
+  const handleLogout = async () => {
+    setAuthLoading(true);
+    try {
+      const url = `${rootUrl}/api/v1/auth/logout`;
+      const response = await axios.get(url);
+      console.log(response);
+      setCurrentUser(null);
+      setAuthLoading(false);
+      return true;
+    } catch (error) {
+      console.error(error);
+      const errorPayload =
+        error instanceof AxiosError ? error.response.data : error;
+      setAuthError(errorPayload);
+      setAuthLoading(false);
+      return false;
+    }
+  };
+
+  return (
+    <UserContext.Provider
+      value={{
+        handleLogin,
+        handleRegister,
+        handleLogout,
+        currentUser,
+        authLoading,
+        authError,
+      }}
+    >
+      {children}
+    </UserContext.Provider>
```

Comment:
We return this component to be rendered. The `values` property is where we stick anything that we want to pass down to components that consume from this context.

`{children}` means anything nested in this component will be rendered as-is.